### ISSUE TITLE
Improve keyboard support in PDF reader

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#1986]
Need to be able to get to annotations and notes via keyboard, close panes, switch tools, etc.

**Comment by mrtcode** [comment]
Any idea how that should work exactly? Are we talking about standard TAB, ENTER, arrow keys navigation or specific keyboard shortcuts as well?

**Comment by bjohas** [comment]
I've built a few keyboard shortcuts into Zutilo - so I suppose I would advocate for making sure that as many features as possible (e.g. hide/show pane, select tool / select colour etc) can be reached via Zutilo. Then it's less critical what is developed directly in the PDF Reader vs. what can be offloaded to Zutilo.

But yes, if the shortcuts can be built into PDF Reader directly, all the better!

**Comment by qqobb** [comment]
Here's a list of pdf.js viewer shortcuts:
https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#faq-shortcuts

**Comment by mrtcode** [comment]
Currently all toolbar buttons, page thumbnails and outline elements can be reached with tab. This is the default PDF.js behaviour.

What we still need:
- Switch between pdf-reader and right-hand side pane (including the split button? sync button?)
- Navigate annotations with tab? (that is currently possible with arrow keys when one of them is selected)
- Navigate links in PDF file (I had to disable that, doesn't seem useful, or is it?)
- Open tags and context menu popup for selected annotation
- Various other shortcuts

Further suggestions are welcome.

**Comment by qqobb** [comment]
> Navigate links in PDF file (I had to disable that, doesn't seem useful, or is it?)

I think this can be very useful for academic PDFs, but only in combination with https://github.com/zotero/zotero/issues/1997. To see this, open [this PDF](https://arxiv.org/pdf/2007.07289.pdf#page=3) in Firefox; Tab, the focus will be on the first reference; Enter, go to the reference; Alt+Left, go back to the original text position (on Windows). It's very convenient for checking all the references without using the mouse.

(BTW, in Zotero 5.0.97-beta.6+ae98488d2, link navigation is not completely switched off. You can get to it with Shift+Tab: Click space to the right of the total page number; press Tab once; after the thumbnail button gets focus, keep pressing Shift+Tab.)

Incidentally, when opening a new PDF, the initial focus should probably be on the page content, such that you can immediately start using the Page Up and Page Down keys.

Let me add some papercuts for the search function. I hope it's ok to mention this here.

- As in Firefox, it would be nice to have Alt+A for "Highlight all", Alt+C for "Match case", Alt+W for "Whole words". These shortcuts should probably only be active when focus is on the search tool. (Feel free to ignore this.)
- Remember the last state for the checkboxes "Highlight all", etc.
- When using the search tool, found text is currently shown too close to the search box. It's not possible to read the found text in context. In Firefox, the scroll position seems to be the same, but you don't have the floating search box. Additional scrolling by the height of the search box could be a solution.

(Sorry, the last two points are off-topic.)

**Comment by dstillman** [comment]
@qqobb: Please post all feature requests and bug reports to the forums.

**Comment by mrtcode** [comment]
The current list of `pdf-reader` keyboard shortcuts. This post will be updated whenever we decide to change something.
We will need to update https://www.zotero.org/support/kb/keyboard_shortcuts at some point.

| Function | Windows/Linux | macOS |
| --- | --- | --- |
| Toggle highlighting tool (*proposal*) | 1 | 1 |
| Toggle note tool (*proposal*) | 2 | 2 |
| Toggle area selection tool (*proposal*) | 3 | 3 |
| Switch color (*proposal*) | 4 | 4 |
| Navigate annotations | <kbd>↑</kbd>, <kbd>↓</kbd> | <kbd>↑</kbd>, <kbd>↓</kbd> |
| Delete annotation | <kbd>Del</kbd>, <kbd>Backspace</kbd> | <kbd>Del</kbd>, <kbd>Backspace</kbd> |
| Deselect annotation, annotation tool, hand tool | <kbd>Escape</kbd> | <kbd>Escape</kbd> |
| Go back | <kbd>Ctrl</kbd> + <kbd>←</kbd>, <kbd>Ctrl</kbd> + <kbd>[</kbd> (*Linux*) | <kbd>Cmd</kbd> + <kbd>[</kbd>, <kbd>Cmd</kbd> + <kbd>←</kbd> |
| Go forward | <kbd>Ctrl</kbd> + <kbd>→</kbd>, <kbd>Ctrl</kbd> + <kbd>]</kbd> (*Linux*) | <kbd>Cmd</kbd> + <kbd>]</kbd>, <kbd>Cmd</kbd> + <kbd>→</kbd> |
| Navigate pages | <kbd>Home</kbd>, <kbd>End</kbd>, <kbd>PgUp</kbd>, <kbd>PgDn</kbd> | <kbd>Home</kbd>, <kbd>End</kbd>, <kbd>PgUp</kbd>, <kbd>PgDn</kbd> |
| Next page | N, J, <kbd>Space</kbd> | N, J, <kbd>Space</kbd> |
| Previous page | P, K, <kbd>Shift</kbd> + <kbd>Space</kbd> | P, K, <kbd>Shift</kbd> + <kbd>Space</kbd> |
| Zoom in | <kbd>Ctrl</kbd> + <kbd>+</kbd>, <kbd>Ctrl</kbd> + <kbd>=</kbd> | <kbd>Cmd</kbd> + <kbd>+</kbd>, <kbd>Cmd</kbd> + <kbd>=</kbd> |
| Zoom out | <kbd>Ctrl</kbd> + <kbd>-</kbd> | <kbd>Cmd</kbd> + <kbd>-</kbd> |
| Zoom auto | <kbd>Ctrl</kbd> + 0 | <kbd>Cmd</kbd> + 0 |
| Rotate the document clockwise | R | R |
| Rotate the document counterclockwise | <kbd>Shift</kbd> + R | <kbd>Shift</kbd> + R |
| Enable the hand tool | H | H |
| Enable the text selection tool ("normal" mode) | S | S |
| Move focus to the 'go to page' box | <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + G | <kbd>Cmd</kbd> + <kbd>Alt</kbd> + G |
| Find text in the document | <kbd>Ctrl</kbd> + F | <kbd>Cmd</kbd> + F |
| Find next occurrence of text in the document | <kbd>Ctrl</kbd> + G | <kbd>Cmd</kbd> + G |
| Print the document | <kbd>Ctrl</kbd> + P | <kbd>Cmd</kbd> + P |

**Comment by dstillman** [comment]
We definitely need to improve keyboard navigation before releasing this. I can't seem to Tab to an annotation in the sidebar at all — it jumps from "Search Annotations" to the toolbar.

I think we'll also want to follow the model (as seen in Firefox, and I think in ARIA guidelines) of Tab jumping between toolbar button groups, with left/right arrow keys moving you between the individual buttons in each group. Here, that's probably 1) the buttons to the left of the page number field, 2) the page number field, 3) the annotation selectors, 4) the Item/Note buttons.

In the annotation list, you should be able to tab to the three-dots button to open the context menu.

Of course, I guess there's the larger issue of how you actually annotate the PDF with just the keyboard. I'm not sure what other apps do there. E.g., can you highlight text in Preview with VoiceOver?

**Comment by qqobb** [comment]
> In the annotation list, you should be able to tab to the three-dots button to open the context menu.

Maybe use Alt instead of Tab to put focus to the three-dots button? In Chrome, pressing Alt once focuses the browser's three-dots button. This also works if you're using Google Docs, where Alt-F selects the File menu in Docs.

Another option might be to enable the context menu button or Shift-F10 as an alternative to right-clicking.

**Comment by mrtcode** [comment]
@dstillman How about dividing into groups like this:
<img width="1364" alt="Screenshot 2022-01-10 at 19 05 05" src="https://user-images.githubusercontent.com/6362989/148808904-da73c860-3c3a-4762-926e-3cc994cc20f7.png">

- Switching focus into/from Item/Notes pane is problematic because of different iframes.
- I'm not sure if toolbar needs smaller groups. [WAI-ARIA example](https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html) and web-library uses relatively large groups. Although Item/Note split button lives outside of pdf-reader iframe which, again, complicates things seriously.
- It might not be easy to implement all this, because all the navigation has to be implemented from zero, and it has to play nice with the existing PDF.js keyboard navigation.

**Comment by mrtcode** [comment]
Ok, here are more details:

**With sidebar:**

<img width="1364" alt="Screenshot 2022-01-11 at 17 39 45" src="https://user-images.githubusercontent.com/6362989/148980626-de31a734-46a0-4f63-9cac-fc812e7555e1.png">

Each number represents Tab key press.

1. Toolbar navigation. ArrowLeft/ArrowRight keys are supported. Item/Notes split button is excluded because it would be too difficult to implement for now. Or maybe it should go after the Nr. 8., when focus escapes from pdf-reader iframe (if escapes). Although switching between pdf-reader and right pane is another unsolved question.
2. ArrowLeft/ArrowRight keys will immediately switch sidebar views, without the need to press Enter.
3. Focus annotations search input. I feel that we need to make that input fixed instead of keeping it inside scrollable annotation container.
4. An annotation will be automatically focused when tabbing into scrollable annotation container. ArrowUp/ArrowDown can be used to navigate between annotations. When focus leaves annotation container, the current annotation will be unselected.
5. Focus three dots button. Enter to open the popup.
6. Focus comment input.
7. Focus tags container. Enter to open tags popup.
8. Focus first color or tag in the selector.

Pressing Tab once again, when number 8 is focused, or pressing Escape will return back focus to PDF page view where you can just scroll or switch pages.

In future we will need to support focusing the first currently visible link in PDF page view (and then navigate with arrow keys).

**Without sidebar:**

<img width="1364" alt="Screenshot 2022-01-11 at 18 00 36" src="https://user-images.githubusercontent.com/6362989/148980652-e164fc88-b29a-40af-9906-d75f0c733c03.png">

It follows basically the same logic.
Tab Nr. 2 will select an annotation (that was last selected before), the same as Tab Nr. 4 in the previous example.

**Comment by dstillman** [comment]
> 4. An annotation will be automatically focused when tabbing into scrollable annotation container.

Does that include expanding the highlight text?

> 5. Focus three dots button. Enter to open the popup.

I think it should be Space, not Enter. (E.g., Firefox toolbar menu)

> 8. Focus first color or tag in the selector.

Space to toggle a tag on/off? Do we have a tag context menu, like in the main tag selector? E.g., to delete a tag from the whole PDF?

I assume there's a (9) here that takes you into the PDF view itself, where you can then use the arrow keys to scroll?

And then without the sidebar, maybe we need (2) for the PDF view itself, where the arrow keys scroll, and (3) jumps to the annotation layer? And it would start with the first annotation in view, not go back to the first one from a previous page.

**Comment by mrtcode** [comment]
> > 4\. An annotation will be automatically focused when tabbing into scrollable annotation container.
> 
> Does that include expanding the highlight text?
> 
> >

No, it will do the same as a single click on an annotation. Enter could be used for highlight expansion.

> I think it should be Space, not Enter. (E.g., Firefox toolbar menu)
> 
> > 8. Focus first color or tag in the selector.
> 
> Space to toggle a tag on/off? Do we have a tag context menu, like in the main tag selector? E.g., to delete a tag from the whole PDF?

Yes, Space. Currently we don't have a context menu to manage tags. I guess we can implement that after the release.



> I assume there's a (9) here that takes you into the PDF view itself, where you can then use the arrow keys to scroll?

Yes, (9) or Escape would always return focus to PDF view. PDF view will always receive focus by default if nothing else is focused or focus is lost. Therefore I don't think that for no-sidebar mode (2) is necessary, because PDF view will already be focused before you start tabbing.

And yes, it should focus the first currently visible annotation.

**Comment by dstillman** [comment]
OK, that all sounds good.

**Comment by ralienpp** [comment]
To this list (https://github.com/zotero/zotero/issues/1986#issuecomment-877064552), I would add `Ctrl+Z`, to undo the last action. For example, If I highlighted the wrong sentence, at the moment I cannot undo the action with the typical hotkey; instead I have to click on the annotation, and press `Del`.

In the same context, I was wondering if it were feasible to customize hotkeys in a more flexible way, giving me the option to set arbitrary keys to the actions I care about (like in this UI for Word):
![image](https://user-images.githubusercontent.com/295338/167506340-1c48f6f8-10a8-4a12-98ab-46658dc19b95.png)

I use the colour of the highlight for different things; ideally, I'd like to be able to press `g` to get green, `y` for yellow, `r` for red, etc. Since not everyone is like me, maybe they'd prefer to use these "premium" keys for other purposes.

**Comment by Zeratoxx** [comment]
+1 here  
https://github.com/zotero/zotero/issues/1986#issuecomment-1121631465

For example, I would like to quickly switch from hand tool to selection tool back and forth, the solution from above would do it for me

**Comment by diogo-rossi** [comment]
Hello every one. Thanks for the great job!

I would like to have the options of "**Zoom to Page Height**" and "**Zoom to Page Width**" also in keyboard shortcuts, to quickly switch between these views.

The solution above will solve for every case, I guess.
